src/xml_parser_main_metadata_together.c

```diff
@@ -336,132 +326,213 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
+        primary_path, filelists_path, other_path, newpkgcb, newpkgcb_data, warningcb, warningcb_data, allow_out_of_order, err
+    );
 
-    cr_CbData cbdata;
-    cbdata.in_progress_pkgs_list = NULL;
-    cbdata.in_progress_pkgs_hash = NULL;
-    cbdata.newpkgcb = newpkgcb;
-    cbdata.newpkgcb_data = newpkgcb_data;
-    cbdata.pkgcb = pkgcb;
-    cbdata.pkgcb_data = pkgcb_data;
+    cr_Package* package = NULL;
+    GError* tmp_err = pkg_iterator->tmp_err;
 
-    if (allow_out_of_order) {
-        cbdata.in_progress_pkgs_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    while (!cr_PkgIterator_is_finished(pkg_iterator)) {
+        int ret = cr_PkgIterator_parseNext(pkg_iterator, &package, err);
+
+        if (ret != CRE_OK) {
+            return ret;
+        }
+        // call user package callback
+        if (pkgcb && pkgcb(package, pkgcb_data, err)) {
+            if (tmp_err)
+                g_propagate_prefixed_error(err, tmp_err, "Parsing interrupted: ");
+            else
+                g_set_error(err, ERR_DOMAIN, CRE_CBINTERRUPTED, "Parsing interrupted");
+            cr_package_free(package);
+            return CR_CB_RET_ERR;
+        } else {
+            // If callback return CRE_OK but it simultaneously set
+            // the tmp_err then it's a programming error.
+            assert(tmp_err == NULL);
+        };
+        cr_package_free(package);
     }
 
+    return cr_PkgIterator_free(pkg_iterator, err);
+}
+
+// tODO: might be possible to select which files to parse by passing NULL as path?
+cr_PkgIterator *
+cr_PkgIterator_new(const char *primary_path,
+                   const char *filelists_path,
+                   const char *other_path,
+                   cr_XmlParserNewPkgCb newpkgcb,
+                   void *newpkgcb_data,
+                   cr_XmlParserWarningCb warningcb,
+                   void *warningcb_data,
+                   gboolean allow_out_of_order,
+                   GError **err)
+{
     assert(primary_path);
     assert(filelists_path);
     assert(other_path);
-    assert(newpkgcb || pkgcb);
     assert(!err || *err == NULL);
 
-    cr_ParserData *primary_pd = NULL;
-    cr_ParserData *filelists_pd = NULL;
-    cr_ParserData *other_pd = NULL;
+    cr_PkgIterator* new_iter = g_new0(cr_PkgIterator, 1);
+    if (!new_iter) {
+        return NULL;
+    }
+
+    new_iter->primary_path = primary_path;
+    new_iter->filelists_path = filelists_path;
+    new_iter->other_path = other_path;
+
+    new_iter->primary_f = NULL;
+    new_iter->filelists_f = NULL;
+    new_iter->other_f = NULL;
+
+    new_iter->primary_pd = NULL;
+    new_iter->filelists_pd = NULL;
+    new_iter->other_pd = NULL;
+
+    new_iter->primary_is_done = 0;
+    new_iter->filelists_is_done = 0;
+    new_iter->other_is_done = 0;
 
-    primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    cr_CbData *cbdata = (cr_CbData*) &new_iter->cbdata;
```

Comment:
I believe this is causing the crash `cbdata` is not actually allocated anywhere, `new_iter` contains only a pointer.
Maybe something like this would work better:
```suggestion
    cr_CbData *cbdata = g_new0(cr_CbData, 1);
    new_iter->cbdata = (void*)cbdata;
```